### ISSUE TITLE
On K8s, EdgeHub does not retains subscriptions after restart (impact modules using azure-iot-sdk-python)

### ISSUE THREAD

**Issue by kiranpradeep** [Azure/iotedge#5292]
## Context
This is a continuation of issue #4050. After #4189, edgeHub retains subscriptions on VMs, but fails to retain on K8s. So raising as new issue.

## Expected Behavior
Python modules on K8s should continue to receive inbound module messages even after edgeHub restart.

## Current Behavior
On K8s, after edgeHub restart, Python edge modules, loses subscriptions. In attached edgeHub logs, kindly find lines `No subscription for receiving messages found for stackedge07-edge/receiver` which indicates missed module messages.

## Steps to Reproduce
1. Sample message sender (every 5 seconds) and message receiver (wait for message) modules using python azure-iot-device sdk [[link](https://github.com/kiranpradeep/bugsquasher/tree/main/iotedge-python-module-losing-subscription)]. The sample is provided only to make sure I am not doing anything bad, otherwise they are ordinary sender/receiver modules. 
2. Restart edgeHub using any mechanism available. (e.g. I set a new environment variable from azure portal on edgeHub). If you have local terminal access [[link](https://docs.microsoft.com/en-us/azure/databox-online/azure-stack-edge-gpu-connect-powershell-interface#use-iotedge-commands)] to K8s cluster, I believe `iotedge restart edgeHub` will restart edgeHub.
3. Receiver module will not get further message from sender

### Device Information
Azure Stack Edge 2106 (GPU)

### Runtime Versions
* aziot-edged: azureiotedge/azureiotedge-iotedged: 1.1.1
* Edge Agent: azureiotedge/azureiotedge-agent: 1.1-k8s-beta14
* Edge Hub: mcr.microsoft.com/azureiotedge-hub:1.2.3

## Logs
Attached [logs.zip](https://github.com/Azure/iotedge/files/6898375/logs.zip)

## Additional Information
I was able to consistently reproduce this issue on Azure Stack Edge and so assuming it would be reproducible on K8s, but I have not tried on K8s directly yet.

**Comment by veyalla** [comment]
The same edgeHub code runs in both environments. Can you check with edgehub v1.1.4?

**Comment by kiranpradeep** [comment]
It is the same code. But doesn't the subscription store location differ? It cannot be in memory.

**Comment by kiranpradeep** [comment]
Same on 1.1.4. Attaching [logs](https://github.com/Azure/iotedge/files/6905453/logs-114.zip) for the same.

**Comment by veyalla** [comment]
Are you setting edge hub's StorageFolder to a persistent volume on the ASE device? This information may help

https://docs.microsoft.com/en-us/azure/iot-edge/how-to-access-host-storage-from-module?view=iotedge-2020-11
https://docs.microsoft.com/en-us/azure/databox-online/azure-stack-edge-gpu-manage-shares

**Comment by kiranpradeep** [comment]
Thanks a lot @veyalla. After setting persistent volume, subscription are retained and issue is no more reproducible.

**Comment by kiranpradeep** [comment]
I wish the [docs](https://docs.microsoft.com/en-us/azure/iot-edge/how-to-access-host-storage-from-module?view=iotedge-2020-11) are updated to reflect this. At the moment, docs are as if edgeHub persistence is for offline scenarios. In ASE/K8s, even when fully online, now it is mandatory even for consistent message reception.

**Comment by veyalla** [comment]
Glad you're unblocked, since this behavior is specific to Kubernetes - it is documented at: https://microsoft.github.io/iotedge-k8s-doc/bp/storage/perstorage.html#configure-edgehub-with-a-pvc-to-avoid-message-loss

**Comment by kiranpradeep** [comment]
@veyalla Requesting to reopen this issue. Kindly find attached [logs](https://github.com/Azure/iotedge/files/6951895/B-logs-from-edgehub-in-edgehub-55f7d74d58-spxg4.txt). If the edgeHub cannot access persistent volume for a previously active lock, it falls back to in-memory store resulting in `azure-iot-device` based Python modules never receiving a message or twin update again. I see this when I restart ASE. If I keep monitoring the kubernetes dashboard at the same, I can see two edgeHubs on two different pods - active (one of which will die soon).

```
Error creating RocksDB store. Falling back to in-memory store.
RocksDbSharp.RocksDbException: IO error: While lock file: /offline_store/edgeHub/LOCK: Resource temporarily unavailable
```

The issue is not consistently reproducible. I have 7 stack edges which I reboot every day in the hope to reproduce. I get this rarely. I have updated the code sample to include persistence on edgeHub. 

<sub>If the user has specified storage, I wish the edgeHub blocks/retry when it cannot access the store,</sub>

**Comment by veyalla** [comment]
Can you try adjusting the deployment strategy to [recreate](https://microsoft.github.io/iotedge-k8s-doc/bp/depstrat.html)?

**Comment by kiranpradeep** [comment]
Thanks again @veyalla

**Comment by kiranpradeep** [comment]
@veyalla I had set deployment strategy as _recreate_ for edgeHub alone, but not for modules based on python sdk as they did not use a shared resource. Kindly find [attached](https://github.com/Azure/iotedge/files/7090233/logs-from-edgehub-in-edgehub-64874ddb5b-4kmx2.txt) debug logs for edgeHub started on 25/Aug. I applied new deployment on 27/Aug which updated python sdk modules (not the edgeHub), during which I have log lines as below.

`<4> 2021-08-27 05:00:55.775 +00:00 [WRN] - Closing connection for device: stackedge09-tm2-edge/DetectionTracer, Microsoft.Azure.Devices.Edge.Hub.Core.MultipleConnectionsException: Multiple connections detected for device stackedge09-tm2-edge/DetectionTracer`

After that, python sdk based modules did not receive further incoming messages as shown below log line

`<4> 2021-08-27 06:23:52.322 +00:00 [WRN] - No subscription for receiving messages found for stackedge09-tm2-edge/DetectionTracer`

It appears like edgeHub destroyed the subscription for python modules during the multiple connections error. Does this mean that *recreate* strategy should be used for all python modules irrespective of whether they use shared resource or not? Kindly suggest.
